daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SValue.scala

```diff
@@ -30,13 +30,99 @@ sealed trait SValue {
   /** Convert a speedy-value to a value which may not be correctly normalized.
     * And so the resulting value should not be serialized.
     */
-  def toUnnormalizedValue: V[V.ContractId] = SValue.toValue(this)
+  def toUnnormalizedValue: V[V.ContractId] = {
+    toValue(
+      disallowGenMapAtVersion = None,
+      eraseType = false,
+    )
+  }
 
   /** Convert a speedy-value to a value normalized according to the LF version.
     */
   def toNormalizedValue(version: TransactionVersion): V[V.ContractId] = {
-    val v = SValue.toValue(this)
-    Util.assertNormalizeValue(v, version) //TODO: avoid separate pass; inline norm into toValue
+    import Ordering.Implicits.infixOrderingOps
+    toValue(
+      disallowGenMapAtVersion =
+        if (version >= TransactionVersion.minGenMap) None else Some(version),
+      eraseType = version >= TransactionVersion.minTypeErasure,
+    )
+  }
+
+  private def toValue(
+      disallowGenMapAtVersion: Option[TransactionVersion],
+      eraseType: Boolean,
+  ): V[V.ContractId] = {
+
+    def maybeEraseTypeInfo[X](x: X): Option[X] =
+      if (eraseType) {
+        None
+      } else {
+        Some(x)
+      }
+
+    def go(v: SValue, maxNesting: Int = V.MAXIMUM_NESTING): V[V.ContractId] = {
+      if (maxNesting < 0)
+        throw SError.SErrorDamlException(interpretation.Error.ValueExceedsMaxNesting)
+      val nextMaxNesting = maxNesting - 1
+      v match {
+        case SInt64(x) => V.ValueInt64(x)
+        case SNumeric(x) => V.ValueNumeric(x)
+        case SText(x) => V.ValueText(x)
+        case STimestamp(x) => V.ValueTimestamp(x)
+        case SParty(x) => V.ValueParty(x)
+        case SBool(x) => V.ValueBool(x)
+        case SUnit => V.ValueUnit
+        case SDate(x) => V.ValueDate(x)
+
+        case SRecord(id, fields, svalues) =>
+          V.ValueRecord(
+            maybeEraseTypeInfo(id),
+            (fields.toSeq zip svalues.asScala)
+              .map { case (fld, sv) => (maybeEraseTypeInfo(fld), go(sv, nextMaxNesting)) }
+              .to(ImmArray),
+          )
+        case SVariant(id, variant, _, sv) =>
+          V.ValueVariant(maybeEraseTypeInfo(id), variant, go(sv, nextMaxNesting))
+        case SEnum(id, constructor, _) =>
+          V.ValueEnum(maybeEraseTypeInfo(id), constructor)
+        case SList(lst) =>
+          V.ValueList(lst.map(go(_, nextMaxNesting)))
+        case SOptional(mbV) =>
+          V.ValueOptional(mbV.map(go(_, nextMaxNesting)))
+        case SMap(true, entries) =>
+          V.ValueTextMap(SortedLookupList(entries.map {
+            case (SText(t), v) => t -> go(v, nextMaxNesting)
+            case (_, _) =>
+              throw SError.SErrorCrash(
+                NameOf.qualifiedNameOfCurrentFunc,
+                "SValue.toValue: TextMap with non text key",
+              )
+          }))
+        case SMap(false, entries) =>
+          // TODO: Is it necessary to check GenMap/version here? Can't it just be assumed?
```

Comment:
afaict it’s only a safe guard. speedy should never produce this for older versions but if we messed up it might catch it. @remyhaemmerle-da cna probably confirm that.

Reply:
Yes, I think we can drop the check.